### ISSUE TITLE
Temporal complexity detection and iterative refinement

### ISSUE THREAD

**Issue by justinshenk** [justinshenk/temporal-awareness#15]
## Overview

Questions involving long-term temporal reasoning may be more "complex" in a way the model recognizes internally. This experiment uses complexity detection to trigger iterative refinement, potentially improving temporal reasoning accuracy.

Builds on prior work in active inference control and complexity-aware prompting.

## Hypothesis

1. Long-horizon questions score higher on complexity classifiers than short-horizon questions
2. Prompting the model to "wait and reconsider" on complex questions improves temporal reasoning
3. Combining complexity scores with temporal probes improves classification accuracy

### Step 1: Classify temporal questions by complexity

Use a prompt complexity classifier (e.g., from Hugging Face) to score questions:

```python
from transformers import pipeline

complexity_classifier = pipeline("text-classification", model="complexity-classifier")

short_term_questions = ["What should I eat for lunch?", "Should I reply to this email now?"]
long_term_questions = ["Should I change careers?", "How should we address climate change?"]

for q in short_term_questions + long_term_questions:
    score = complexity_classifier(q)
    print(f"{q[:40]}... -> {score}")
```

**Prediction**: Long-horizon questions score higher complexity.

### Step 2: Complexity-triggered re-evaluation

For high-complexity questions, prompt the model to reconsider:

```python
def iterative_temporal_reasoning(model, question, complexity_threshold=0.7):
    complexity = get_complexity_score(question)
    
    # First pass
    response_1 = model.generate(question)
    activations_1 = extract_activations(model, question)
    temporal_pred_1 = temporal_probe.predict(activations_1)
    
    if complexity > complexity_threshold:
        # Trigger re-evaluation
        prompt_2 = f"""
        {question}
        
        This is a complex question. Take a moment to consider the time horizons involved.
        Think about both immediate and long-term implications before answering.
        """
        response_2 = model.generate(prompt_2)
        activations_2 = extract_activations(model, prompt_2)
        temporal_pred_2 = temporal_probe.predict(activations_2)
        
        return {
            'complexity': complexity,
            'temporal_before': temporal_pred_1,
            'temporal_after': temporal_pred_2,
            'shift': temporal_pred_2 - temporal_pred_1
        }
    
    return {'complexity': complexity, 'temporal': temporal_pred_1}
```

**Question**: Does iterative refinement shift the model toward more appropriate temporal reasoning?

### Step 3: Active inference framing

Interpret the "wait" intervention through active inference:

- Re-iteration as reducing expected free energy
- Model gains information before committing to response
- Track whether re-iteration improves calibration (confidence aligns with accuracy)

```python
def measure_calibration_improvement(results):
    """Compare calibration before/after re-evaluation."""
    before_calibration = compute_ece(results['temporal_before'], results['ground_truth'])
    after_calibration = compute_ece(results['temporal_after'], results['ground_truth'])
    return before_calibration - after_calibration  # Positive = improvement
```

### Step 4: XGBoost ensemble

Combine temporal probe predictions with complexity scores:

```python
import xgboost as xgb

# Features: probe predictions + complexity score
X = np.column_stack([
    temporal_probe_predictions,
    complexity_scores,
    # Optional: layer-wise probe confidences
])
y = ground_truth_temporal_labels

model = xgb.XGBClassifier()
model.fit(X_train, y_train)

# Compare to probe-only baseline
probe_only_acc = accuracy_score(y_test, temporal_probe.predict(X_test_activations))
ensemble_acc = accuracy_score(y_test, model.predict(X_test))
print(f"Probe only: {probe_only_acc:.2%}, Ensemble: {ensemble_acc:.2%}")
```

**Question**: Does the ensemble outperform probes alone?

## Research Questions

1. Are long-horizon questions systematically more "complex" by standard metrics?
2. Does prompting for re-evaluation improve temporal reasoning accuracy?
3. Can we predict when re-evaluation will help vs. hurt?
4. What's the optimal way to combine complexity and temporal probe signals?

## Connection to Other Issues

- **#11 Active inference**: Re-evaluation as free energy minimization
- **#6 Uncertainty calibration**: Does iteration improve calibration?
- **#12 Defabrication**: Does iteration cause temporal representation collapse or refinement?
- **#7 Consistency**: Are iterative responses more temporally consistent?

## Success Criteria

| Metric | Target |
|--------|--------|
| Complexity-horizon correlation | r > 0.3 |
| Calibration improvement from re-evaluation | ECE reduction > 10% |
| Ensemble accuracy lift over probe alone | > 5% absolute |

## Resources

- Prompt complexity classifiers on Hugging Face
- Existing temporal probes from probe validation
- XGBoost for ensemble
